### ISSUE TITLE
Execlog parser should output each action's digest 

### ISSUE THREAD

**Issue by tbaing** [bazelbuild/bazel#16741]
### Description of the feature request:

While trying to determine why a given target results in a remote cache miss, I used the execlog parser to gather detailed logs about what was happening. I wanted to get the action digest for the problematic action, to check for the existence of that object in the remote cache, but didn't see anything I recognized as that. The output contains hashes for each individual input and for the output, but nothing that appeared to be the digest for the aggregate set of all inputs, env variables, command, etc.

Is there some option for adding this information to the output that's not documented on https://bazel.build/remote/cache-remote? If not, can we please add that information to the output from the execlog parser?

### What underlying problem are you trying to solve with this feature?

It's difficult to correlate the output from the execlog parser with the artifacts in the cache, which makes it harder to be certain whether a cache miss is due to the artifact not being present or is due to something else going wrong.

### Which operating system are you running Bazel on?

gLinux

### What is the output of `bazel info release`?

release 6.0.0-pre.20221012.2

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
fatal: not a git repository (or any of the parent directories): .git
```

### Have you found anything relevant by searching the web?

Nothing relevant found.

**Comment by tjgq** [comment]
You're looking for the [`SpawnExec.digest`](https://cs.opensource.google/bazel/bazel/+/master:src/main/protobuf/spawn.proto;l=183;drc=b4b8b2614ae854651075506666f109d0fc508ad1
) field. This is a new addition in Bazel 6.

On a separate note: you might want to use `--execution_log_json_file` to produce the log in JSON format instead of binary and avoid using the execlog parser.

**Comment by tbaing** [comment]
I switched to the JSON format (thanks for the suggestion!) and it's easier to read. I do see the `SpawnExec.digest` field there, but I'm noticing it's not present in some cases like local execution, which is exactly the situation where I'd want to see its value. Is there a reason it's not present for local execution?

**Comment by tjgq** [comment]
You are correct - the field is missing for locally executed actions because the corresponding data structure ([`RemoteCacheClient.ActionKey`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/common/RemoteCacheClient.java;l=39;drc=d1a9f156135d6e002e06929c06b6cac6650fa8e2)) is only created for actions that run remotely.

I agree that producing a digest for locally executed actions would be helpful, but I'm not sure about the best way to do it; creating an `ActionKey` for an action that won't run remotely might be unnecessarily expensive.

**Comment by tbaing** [comment]
For the action to run locally, wouldn't we have had to compute the ActionKey and use it to check the remote cache, and only execute locally after a cache miss? Shouldn't the ActionKey already exist by this point?

**Comment by tjgq** [comment]
There's also the question of whether/how we should populate the field if neither remote caching nor remote execution are enabled.

**Comment by tbaing** [comment]
The difficulty of "we have it but not here" makes sense.

Maybe alternatively we could include information about the cache interactions (checking and writing) in the raw log for the run? That would allow us to see the ActionKey being used and the success/failure results for each interaction with the cache, eliminating both the problem of not having the necessary info at the time we construct the content for the actual execution and the problem of what to do if no remote cache interaction is occurring.

**Comment by tjgq** [comment]
@tbaing Bazel has its own concept of an "action key" [1] which every action has, independently of the strategy it ends up running under. What do you think about adding it to a separate `SpawnExec.action_key` field? This would let you correlate invocations of "the same action" across invocations; the drawback is that there's no way to convert it to and from the Remote Execution key (but it's not clear to me that's a requirement).

[1] https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/actions/ActionAnalysisMetadata.java;l=98;drc=14e32e7f5d1492f7ca4ffc17c40b913ec5799603

**Comment by tjgq** [comment]
> Would the remote cache key be 1:1 with that internal action key, even though not identical/computable from one another? Or does one of them consider information that the other does not, potentially resulting in multiple values on one side being mapped to a single value on the other?

It's definitely possible for two internally distinct actions to map to the same remote cache entry. I don't believe the reverse is possible.

**Comment by brentleyjones** [comment]
@bazel-io flag

**Comment by keertk** [comment]
@bazel-io fork 7.0.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
